### PR TITLE
ARROW-7205 : [C++][Gandiva] Implement regexp_like in Gandiva

### PR DESCRIPTION
**Pull request by wjones127**
Was looking for "rlike" support in Gandiva and found we had nearly implementing it. This PR revives apache/arrow#5860.

### CODE REVIEW

**Comment by wjones127** [comment]
@pravindra @Praveen2112 @projjal Got some time for a review? 😁

**Comment by projjal** [comment]
Hi @wjones127. Looks like creating a base holder class and two derived class for just these two functions seems overkill to me (I think thats why I closed the original pr temporarily). The only difference in logic between the two functions is that in case of "like" we convert the sql pattern to regex pattern before storing the compiled pattern. Would it make sense if we pass a flag to the holder that says if this is sql pattern or regex patttern?

**Comment by wjones127** [comment]
Projjal, that's a good point. The SQL like function is basically a wrapper around `regexp_like`. I don't know why I didn't see the optimizations should be the same.

I will consolidate them into a single `LikeHolder`.

File: cpp/src/gandiva/like_holder.cc [hunk 1]
```diff
@@ -62,39 +56,95 @@ const FunctionNode LikeHolder::TryOptimize(const FunctionNode& node) {
...
 
-  return Make(arrow::util::get<std::string>(literal->holder()), holder);
+  return Status::OK();
 }
 
-Status LikeHolder::Make(const std::string& sql_pattern,
-                        std::shared_ptr<LikeHolder>* holder) {
-  std::string pcre_pattern;
-  ARROW_RETURN_NOT_OK(RegexUtil::SqlLikePatternToPcre(sql_pattern, pcre_pattern));
+Result<std::string> RegexpMatchesHolder::GetPattern(const FunctionNode& node) {
```
Comment (wjones127) [inline cpp/src/gandiva/like_holder.cc:105]: FYI This looks to be the first use of `Arrow::Result` in the Gandiva codebase, but this seems to be what the main arrow C++ codebase is using recently. Let me know if you object.

File: cpp/src/gandiva/like_holder.cc [hunk 2]
```diff
@@ -62,39 +56,95 @@ const FunctionNode LikeHolder::TryOptimize(const FunctionNode& node) {
   return node;
 }
 
+const FunctionNode SQLLikeHolder::TryOptimize(const FunctionNode& node) {
```
Comment (wjones127) [inline cpp/src/gandiva/like_holder.cc:59]: `like` will now always optimize into `regexp_matches` or one of the optimizations above. I found this to be the easiest way to reuse the optimization code.

File: cpp/src/gandiva/regex_util.h [hunk 3]
```diff
@@ -29,7 +29,7 @@ namespace gandiva {
 /// \brief Utility class for converting sql patterns to pcre patterns.
 class GANDIVA_EXPORT RegexUtil {
  public:
-  // Convert an sql pattern to a pcre pattern
+  // Convert an sql pattern to a pcre pattern for use with PartialMatch
```
Comment (wjones127) [inline cpp/src/gandiva/regex_util.h:32]: Previously, `like` was using `RE2::FullMatch`. However, `regexp_matches` needs to use `RE2::PartialMatch`, so I've modified this method to make partial match statements. I have added a new set of tests to validate it.

I didn't see any use of this utility method outside of the `like` implementation, so I thought this change would be okay.

File: cpp/src/gandiva/regex_util.cc [hunk 4]
```diff
@@ -30,8 +30,14 @@ Status RegexUtil::SqlLikePatternToPcre(const std::string& sql_pattern, char esca
...
 
+    if (idx == 0 && cur != '%') {
+      pcre_pattern += '^';
+    } else if (idx == 0 && cur == '%') {
+      continue;
+    }
+
     // Escape any char that is special for pcre regex
-    if (pcre_regex_specials_.find(cur) != pcre_regex_specials_.end()) {
+    if (pcre_regex_specials_.find(cur) != pcre_regex_specials_.end() && cur != escape_char) {
```
Comment (wjones127) [inline cpp/src/gandiva/regex_util.cc:40]: This change enables using `\` as an escape character in LIKE statements. For example, `col like '\_%'` can match anything starting with an underscore.

**Comment by wjones127** [comment]
I think we should merge #9700 before this one, since my refactor of `LikeHolder` would cause merge conflicts on that PR. So I'll wait for that merge and then finish this.

**Comment by nealrichardson** [comment]
> I think we should merge #9700 before this one, since my refactor of `LikeHolder` would cause merge conflicts on that PR. So I'll wait for that merge and then finish this.

@wjones127 #9700 has merged, are you planning to pick this back up?

**Comment by wjones127** [comment]
> 
> 
> > I think we should merge #9700 before this one, since my refactor of `LikeHolder` would cause merge conflicts on that PR. So I'll wait for that merge and then finish this.
> 
> @wjones127 #9700 has merged, are you planning to pick this back up?

Yes, I'll work on it this weekend.

**Comment by wjones127** [comment]
@projjal This is now ready for review.

Since we've added `ilike` and special escape_char options for `like`, I made `SQLLikeHolder` inherit from `RegexMatchHolder` instead of sharing the same holder class. I did make sure the `TryOptimize` logic wasn't duplicated.

**Pull request closed without merging**